clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -449,16 +465,29 @@ private static int parseAcks(String acksString) {
             if (callback != null)
                 callback.onCompletion(null, e);
             this.errors.record();
+            if (this.interceptors != null)
+                this.interceptors.onAcknowledgement(null, e);
             return new FutureFailure(e);
         } catch (InterruptedException e) {
             this.errors.record();
+            if (this.interceptors != null)
+                this.interceptors.onAcknowledgement(null, e);
             throw new InterruptException(e);
         } catch (BufferExhaustedException e) {
             this.errors.record();
             this.metrics.sensor("buffer-exhausted-records").record();
+            if (this.interceptors != null)
+                this.interceptors.onAcknowledgement(null, e);
             throw e;
         } catch (KafkaException e) {
             this.errors.record();
+            if (this.interceptors != null)
+                this.interceptors.onAcknowledgement(null, e);
+            throw e;
+        } catch (Exception e) {
```

Comment:
I wonder if it would not be better to have a `catch (Exception e)` that does both the recording of errors and calls `onAcknowledgement` followed by a series of `isinstanceof` checks (in Scala, this would be a bit nicer as pattern matching for catching exceptions is the same as standard pattern matching, but oh well).

Duplicated logic like this inevitably leads to bugs.